### DISCUSSION TITLE
Source Generators and Source Control

### DISCUSSION THREAD

**Discussion by Grauenwolf** [dotnet/csharplang#4655] [Language Ideas]
There should be a way to allow code emitted by a Source Generator to be checked into source control. 

Generated code can often be very complex. And if something breaks, the developer will need to be able to see the current and previous state of the code in order to determine what changed.

This should be an optional setting, as some people would prefer not to have generated code in source control.

Placing the generated source code in the solution tree will also make it easier to set break points in that code when debugging. (Currently this has to be done by stepping into the code, and then setting a break point in the temp file.)

**Comment by Grauenwolf** [comment]
Looks like this was an intended feature, but I can't find any reference to it being implemented.

> To support the use case where a user wishes to generate the source text, then commit the generated files to source control, we will allow changing the location of the generated files via an appropriate command line switch, and matching MSBuild property (naming still to be determined).

https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.md

**Comment by 333fred** [comment]
You'll want to set `CompilerGeneratedFilesOutputPath` to some non-null value, and exclude the generated files from compilation.

Example usage:
https://github.com/dotnet/roslyn/blob/315c2e149ba7889b0937d872274c33fcbfe9af5f/src/Compilers/CSharp/Portable/Microsoft.CodeAnalysis.CSharp.csproj#L21
https://github.com/dotnet/roslyn/blob/315c2e149ba7889b0937d872274c33fcbfe9af5f/src/Compilers/CSharp/Portable/Microsoft.CodeAnalysis.CSharp.csproj#L49-L52

**Reply by 333fred** [reply]
Also @chsienki to update the source generators doc with that.

**Comment by Grauenwolf** [comment]
Necessary project changes:

1. Add `EmitCompilerGeneratedFiles` property
2. Add `CompilerGeneratedFilesOutputPath` property
3. Exclude the generated folder. Otherwise the compiler will try to add the same file twice (one from disk, one from the code generator) and you'll get compiler errors.


  <PropertyGroup>
    <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
    <CompilerGeneratedFilesOutputPath>Generated</CompilerGeneratedFilesOutputPath>
  </PropertyGroup>

  <ItemGroup>
    
    <Compile Remove="$(CompilerGeneratedFilesOutputPath)/*/**/*.cs" />
  </ItemGroup>
